# Homework 3 : Functions, Histograms, and Groups

**Reading**: 

* [Visualizing Numerical Distributions](hhttps://ipydatabook.github.io//chapters/07/2/visualizing-numerical-distributions.html) 
* [Functions and Tables](https://ipydatabook.github.io//chapters/08/functions-and-tables.html)

Please complete this notebook by filling in the cells provided.

**Throughout this homework and all future ones, please be sure to not re-assign variables throughout the notebook!** For example, if you use `max_temperature` in your answer to one question, do not reassign it later on. Moreover, please be sure to only put your written answers in the provided cells. 

In [16]:
# Don't change this cell; just run it. 

import numpy as np
from datascience import *

import scipy.stats as stats

# These lines do some fancy plotting magic.\n",
import matplotlib
%matplotlib inline
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')

import warnings
warnings.filterwarnings("ignore", category=np.VisibleDeprecationWarning) 

warnings.filterwarnings("ignore", category=UserWarning) 


Before continuing the assignment, select "Save and Checkpoint" in the File menu.

## Burrito-ful San Diego

Tam, Margaret and Winifred are trying to use Data Science to find the best burritos in San Diego! Their friends Irene and Maya provided them with two comprehensive datasets on many burrito establishments in the San Diego area taken from (and cleaned from): https://www.kaggle.com/srcole/burritos-in-san-diego/data

The following cell reads in a table called `ratings` which contains names of burrito restaurants, their Yelp rating, Google rating, as well as their Overall rating. It also reads in a table called `burritos_types` which contains names of burrito restaurants, their menu items, and the cost of the respective menu item at the restaurant.

**Question 1.** Use Table.read_table() to read in a data set called "Burritos.csv" and name it burritos.

*Hint: If you need refreshers on table methods, look at the [python reference](http://datasci.pythonanywhere.com/pythonreference).*

<!--
BEGIN QUESTION
name: q1_1
-->

In [3]:
burritos = ...
burritos.show(5)

AttributeError: 'ellipsis' object has no attribute 'show'

**Question 2.** Let's look at how the Yelp scores compare to the Google scores in the `burritos` table. First, assign `yelp_and_google` to a table only containing the columns `Yelp` and `Google`. Then, make a scatter plot with Yelp scores on the x-axis and the Google scores on the y-axis. 

<!--
BEGIN QUESTION
name: q1_2
manual: True
-->
<!-- EXPORT TO PDF -->

In [ ]:
yelp_and_google = ...
...

# Don't change/edit/remove the following line.
# To help you make conclusions, we have plotted a straight line on the graph (y=x)
plt.plot(np.arange(2.5,5,.5), np.arange(2.5,5,.5));

**Question 3.** Looking at the scatter plot you just made in Question 1.2, do you notice any pattern(s) (i.e. is one of the two types of scores consistently higher than the other one)? If so, describe them **briefly** in the cell below.

<!--
BEGIN QUESTION
name: q1_3
manual: True
-->
<!-- EXPORT TO PDF -->

*Write your answer here, replacing this text.*

Here's a refresher on how `.group` works! You can read how `.group` works in the [textbook](https://ipydatabook.github.io//chapters/08/2/Classifying_by_One_Variable.html), or you can view the video below. The video resource was made by a past staff member - Divyesh Chotai!

In [ ]:
from IPython.display import YouTubeVideo
YouTubeVideo("HLoYTCUP0fc")

**Question 4.** From the `burritos` table, some of the restaurant locations have multiple reviews. Winifred thinks California burritos are the best type of burritos, and wants to see the average overall rating for California burritos at each location. Create a table that has two columns: the name of the restaurant and the average overall rating of California burritos at each location.

*Tip: Revisit the burritos table to see how California burritos are represented.*

*Note: you can break up the solution into multiple lines, as long as you assign the final output table to `california_burritos`! For reference however, the staff solution only used one line.*

<!--
BEGIN QUESTION
name: q1_4
-->

In [ ]:
california_burritos = ...
california_burritos

**Question 5.** Given this new table `california_burritos`, Winifred can figure out the name of the restaurant with the highest overall average rating! Assign `best_restaurant` to a line of code that evaluates to a string that corresponds to the name of the restaurant with the highest overall average rating. 

<!--
BEGIN QUESTION
name: q1_5
-->

In [ ]:
best_restaurant = ...
best_restaurant

**Question 6.** Using the `burritos` table, assign `menu_average` to a table that has three columns that uniquely pairs the name of the restaurant, the menu item featured in the review, and the average Overall score for that menu item at that restaurant.

*Hint: Use .group, and remember that you can group by multiple columns. Here's an example from the [textbook](https://www.inferentialthinking.com/chapters/08/3/Cross-Classifying_by_More_than_One_Variable.html)*.

<!--
BEGIN QUESTION
name: q1_6
-->

In [ ]:
menu_average = ...
menu_average

**Question 7.** Tam thinks that burritos in San Diego are cheaper (and taste better) than the burritos in Berkeley. Plot a histogram that visualizes that distribution of the costs of the burritos from San Diego in the `burritos` table. Also use the provided `bins` variable when making your histogram, so that visually the histogram is more informative.

<!--
BEGIN QUESTION
name: q1_7
manual: True
-->
<!-- EXPORT TO PDF -->

In [ ]:
bins = np.arange(0, 15, 1)
# Please also use the provided bins

...


## 2. Faculty salaries


This exercise is designed to give you practice using the Table methods `pivot` and `group`. [Here](http://datasci.pythonanywhere.com/pythonreference/) is a link to the Python reference page in case you need a quick refresher.

Run the cell below to view a demo on how you can use pivot on a table. (Thank you to past staff Divyesh Chotai)

In [ ]:
from IPython.display import YouTubeVideo
YouTubeVideo("4WzXo8eKLAg")

In the next cell, we load a dataset created by the [Daily Cal](http://projects.dailycal.org/paychecker/) which contains Berkeley faculty, their departments, their positions, and their gross salaries in 2015.

In [9]:
raw_profs = Table.read_table("faculty.csv").where("year", are.equal_to(2015)).drop("year", "title")
profs = raw_profs.relabeled("title_category", "rank")
profs

name            | department                      | rank                | gross_salary
CYNTHIA ABAN    | South & Southeast Asian Studies | lecturer            | 64450
PIETER ABBEEL   | Computer Science                | associate professor | 184998
SALLY ABEL      | Law                             | lecturer            | 3466
ELIZABETH ABEL  | English                         | professor           | 138775
DOR ABRAHAMSON  | Education                       | associate professor | 100300
KATHRYN ABRAMS  | Law                             | professor           | 319693
BARBARA ABRAMS  | Public Health                   | professor           | 191162
SARAH ACCOMAZZO | Social Welfare                  | lecturer            | 14779
CHARISMA ACEY   | City and Regional Planning      | assistant professor | 101567
DAVID ACKERLY   | Biology                         | professor           | 182288
... (2049 rows omitted)

We want to use this table to generate arrays with the names of each professor in each department.

**Question 1.** For each rank, who is the highest paid person in this data set?  Use the ```group``` function and the ```max``` function to create a small table showing the highest paid faculty member within each rank.  

<!--
BEGIN QUESTION
name: q2_1
-->

In [8]:
#Put your code here

profs.group("rank", max)

position            | name max             | department max                         | gross_salary max
assistant professor | ZACHARY PARDOS       | Theater, Dance and Performance Studies | 312211
associate professor | ZSOLT KATONA         | Theater, Dance and Performance Studies | 337764
lecturer            | ZEBA NAZEERI-SIMMONS | Theater, Dance and Performance Studies | 273800
professor           | ZUO-JUN SHEN         | Theater, Dance and Performance Studies | 507589

rank                | name mean | department mean | gross_salary mean
assistant professor |           |                 | 131831
associate professor |           |                 | 128869
lecturer            |           |                 | 38190.3
professor           |           |                 | 199949

**Question 2.** Using a combination of `group`, `select`, `np.mean`, and `relabel`, create this table.

|Rank               | Average Salary|
|-------------------|---------------|
|assistant professor|131831         |
|associate professor|128869         |
|lecturer           |38190.3        |
|professor          |199949         |



In [ ]:
#Put your code here



**Question 3.** Set `department_ranges` to a table containing departments as the rows, and the position as the columns. The values in the rows should correspond to a salary range, where range is defined as the **difference between the highest salary and the lowest salary in the department for that position**. 

*Hint:* First we will define a custom function `range` which takes in an array and returns the range (difference between minimum and maximum) of the numbers in that array. 

<!--
BEGIN QUESTION
name: q2_3
manual: false
-->

In [17]:
# Define a custom range first
def range(array):
    '''Computes diff between max and min in array'''
    diff = np.max(array) - np.min(array)
    return diff

department_ranges = profs.select('department', 'rank', 'gross_salary').pivot('department','rank', 'gross_salary', collect = range)
department_ranges




rank                | African American Studies | Agricultural and Resource Economics and Policy | Anthroplogy | Architecture | Art History | Art Practice | Astronomy | Bioengineering | Biology | Buddhist Studies | Business | Chemical Engineering | Chemistry | City and Regional Planning | Civil and Environmental Engineering | Classics | College Writing | Comparative Literature | Computer Science | Demography | Earth and Planetary Science | East Asian Languages and Culture | East Asian Studies | Economics | Education | Electrical Engineering | Electrical Engineering and Computer Science | Energy and Resources Engineering | English | Environmental Design | Environmental Science, Policy and Management | Ethnic Studies | Film and Media | French | Gender and Women's Studies | Geography | German | History | Industrial Engineering & Operations Research | Information | Interdisciplinary Studies | International and Area Studies | Italian | Journalism | Landscape Architecture & Environmental Planning | Law    | Linguistics | Material Science and Engineering | Mathematics | Mechanical Engineering | Middle Eastern Studies | Music  | Near Eastern Studies | Nuclear Engineering | Nutritional science | Optometry | Philosophy | Physics | Political Science | Psychology | Public Health | Public Policy | Rhetoric | Scandinavian Languages | Slavic Languages and Literature | Social Welfare | Sociology | South & Southeast Asian Studies | Spanish and Portuguese | Statistics | Theater, Dance and Performance Studies
assistant professor | 128828                   | 0                                              | 11193       | 47675        | 57288       | 0            | 0         | 5513           | 209667  | 0                | 217211   | 50004                | 171193    | 6433                       | 88702                               | 0        | 0               | 0                      | 150924           | 0          | 26181                       | 45517                            | 0                  | 174568    | 74228     | 24445                  | 0                                           | 79175                            | 107475  | 0                    | 0                                            | 107327         | 48050          | 0      | 0                          | 0         | 0      | 100035  | 67968                                        | 0           | 0                         | 0                              | 0       | 0          | 3702                                            | 0      | 98519       | 77030                            | 64657       | 81168                  | 0                      | 0      | 18766                | 16357               | 0                   | 0         | 3822       | 114303  | 69015             | 10467      | 75303         | 167195        | 58735    | 44325                  | 0                               | 17576          | 6941      | 8809                            | 19509                  | 20474      | 0
associate professor | 48814                    | 0                                              | 103271      | 103204       | 27439       | 26632        | 0         | 2794           | 113341  | 0                | 258216   | 0                    | 0         | 30339                      | 33286                               | 38342    | 0               | 57587                  | 56559            | 0          | 4798                        | 61196                            | 0                  | 150638    | 50667     | 47038                  | 0                                           | 0                                | 103961  | 0                    | 3530                                         | 41395          | 0              | 14484  | 12986                      | 61472     | 36237  | 104938  | 0                                            | 145575      | 0                         | 0                              | 61938   | 0          | 41677                                           | 

**Question 4.** Give an explanation as to why some of the row values are `0` in the `department_ranges` table from the previous question.

<!--
BEGIN QUESTION
name: q2_3
manual: True
-->
<!-- EXPORT TO PDF -->

*Write your answer here, replacing this text.*

## 3.  Which department gets paid the most?

This question is vague, when you really know statistics.  

It could mean, "Which department has the highest average salary?" or even "median salary?"

It could mean, "Which department has the highest maximum salary?"

It could also mean, "Which department has the highest sum of salaries?"

Why don't we check them all.  At the same time, should we discuss what these different statistics might tell us and under what circumstances they would matter?

In the cells below, enter the code necessary to answer these questions.

**Question 3.1** Which department has the highest average salary?

**Question 3.2** Which department has the highest median salary?

**Question 3.3** Which department has the highest maximum salary?

**Question 3.4** Which department has the highest sum of salaries?


For each of these, use code to create table with 1 row containing the name of department and the actual value (mean, median, maximum, or sum) requested.  

In [ ]:
## Use the rest of this cell to answer 3.1 



In [ ]:
## Use the rest of this cell to answer 3.2 



In [ ]:
## Use the rest of this cell to answer 3.3 



In [20]:
## Use the rest of this cell to answer 3.4 

profs.select("department", "gross_salary").group("department", sum).relabel(1,"Total Salaries").sort("Total Salaries", descending = True).take(0)

department | Total Salaries
Business   | 2.6727e+07